src/miniscript/satisfy.rs

```diff
@@ -212,339 +238,521 @@ impl_tuple_satisfier!(&'a A, &'b B, &'c C, &'d D, &'e E, &'f F,);
+                        stack: Witness::combine(r_sat.stack, Witness::push_0()),
+                        has_sig: r_sat.has_sig,
+                    },
+                )
+            },
             Terminal::Thresh(k, ref subs) => {
-                fn flatten(v: Vec<Vec<Vec<u8>>>) -> Vec<Vec<u8>> {
-                    v.into_iter().fold(vec![], |mut acc, x| {
-                        acc.extend(x);
-                        acc
-                    })
-                }
+                let mut sats = subs
+                    .iter()
+                    .map(|s| Self::satisfy(&s.node, stfr))
+                    .collect::<Vec<_>>();
+                // Start with the to-return stack set to all dissatisfactions
+                let mut ret_stack = subs
+                    .iter()
+                    .map(|s| Self::satisfy(&s.node, stfr))
+                    .collect::<Vec<_>>();
+
+                // Sort everything by (sat cost - dissat cost), except that
+                // satisfactions without signatures beat satisfactions with
+                // signatures
+                let mut sat_indices = (0..subs.len()).collect::<Vec<_>>();
+                sat_indices.sort_by_key(|&i| {
+                    let stack_weight = match (&sats[i].stack, &ret_stack[i].stack) {
+                        (&Witness::Unavailable, _) => i64::MAX,
+                        (_, &Witness::Unavailable) => i64::MIN,
+                        (&Witness::Stack(ref s), &Witness::Stack(ref d)) => {
+                            s.iter().map(Vec::len).sum::<usize>() as i64
+                                - d.iter().map(Vec::len).sum::<usize>() as i64
+                        }
+                    };
+                    (sats[i].has_sig, stack_weight)
+                });
 
-                if k == 0 {
-                    return Some(vec![]);
+                for i in 0..k {
+                    mem::swap(&mut ret_stack[i], &mut sats[i]);
                 }
 
-                let mut satisfied = 0;
-                let mut ret = Vec::with_capacity(subs.len());
-                let mut ret_dis = Vec::with_capacity(subs.len());
-
-                for sub in subs.iter().rev() {
-                    let dissat = sub.node.dissatisfy(satisfier)?;
-                    if let Some(sat) = sub.node.satisfy(satisfier, age, height) {
-                        ret.push(sat);
-                        satisfied += 1;
-                    } else {
-                        ret.push(dissat.clone());
+                // The above loop should have taken everything without a sig
+                // (since those were sorted higher than non-sigs). If there
+                // are remaining non-sig satisfactions this indicates a
+                // malleability vector
+                if k < sats.len() && !sats[k].has_sig {
+                    Satisfaction {
+                        stack: Witness::Unavailable,
+                        has_sig: false,
```

Comment:
Should this technically be `ret_stack.iter().any(|sat| sat.has_sig),`? maybe false also works

Reply:
I could add an assertion that `ret_stack.iter().any(|sat| sat.has_sig)` is false; none of the dissatisfactions should have signatures (because dissatisfactions of `d` fragments never do) while none of the satisfactions should have signatures either (because if this branch is being triggered, they're all weak).

Reply:
Added an assertion.